### PR TITLE
Error on nested impl Trait and path projections from impl Trait

### PR DESCRIPTION
**Pull request by cramertj**
cc #34511

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_passes/ast_validation.rs [hunk 1]
```diff
@@ -420,6 +420,141 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
...
+    fn visit_path_parameters(&mut self, _: Span, path_parameters: &'a PathParameters) {
+        match *path_parameters {
+            PathParameters::AngleBracketed(ref params) => {
+                for type_ in &params.types {
+                    self.visit_ty(type_);
+                }
+                for type_binding in &params.bindings {
+                    // Type bindings such as `Item=impl Debug` in `Iterator<Item=Debug>`
+                    // are allowed to contain nested `impl Trait`.
+                    self.with_impl_trait(None, |this| visit::walk_ty(this, &type_binding.ty));
```
Comment (nikomatsakis) [inline src/librustc_passes/ast_validation.rs]: I think this code would permit this

```
<dyn Iterator<Item = impl Debug> as Iterator>::Item
```

which ought not to be allowed

Comment (nikomatsakis) [reply src/librustc_passes/ast_validation.rs]: Oh, probably not, because that is handled by a totally distinct visitor, I see

Comment (nikomatsakis) [reply src/librustc_passes/ast_validation.rs]: Can you add a test for it, anyway?

File: src/librustc_passes/ast_validation.rs [hunk 2]
```diff
@@ -420,6 +420,141 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
...
+
+
+impl<'a> Visitor<'a> for NestedImplTraitVisitor<'a> {
+    fn visit_ty(&mut self, t: &'a Ty) {
+        if let TyKind::ImplTrait(_) = t.node {
+            if let Some(outer_impl_trait) = self.outer_impl_trait {
+                struct_span_err!(self.session, t.span, E0666,
+                                 "nested `impl Trait` is not allowed")
+                    .span_label(outer_impl_trait, "outer `impl Trait`")
+                    .span_label(t.span, "devilishly nested `impl Trait` here")
```
Comment (nikomatsakis) [inline src/librustc_passes/ast_validation.rs]: I don't "devilishly" should be in our error message

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
New test, new error message, but seems good

File: src/librustc_passes/ast_validation.rs (hunk 2 above)
Comment (cramertj) [reply src/librustc_passes/ast_validation.rs]: No love for the E0666 easter egg :cry:

Comment (cramertj) [reply src/librustc_passes/ast_validation.rs]: Done.

File: src/librustc_passes/ast_validation.rs (hunk 1 above)
Comment (cramertj) [reply src/librustc_passes/ast_validation.rs]: Done.

**Comment by nikomatsakis** [comment]
Travis is unhappy:

https://travis-ci.org/rust-lang/rust/builds/341238329#L1665-L1686

The `ui/error-codes/E0657.rs` test is failing.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
r=me once travis is happy

**Pull request merged**